libs/test_positivity.py

```diff
@@ -33,65 +49,60 @@ def calculate(self, delta_df: pd.DataFrame) -> pd.DataFrame:
 
 
 TEST_POSITIVITY_METHODS = (
-    Method(
-        "positiveCasesViral_totalTestEncountersViral",
-        CommonFields.POSITIVE_CASES_VIRAL,
-        CommonFields.TOTAL_TEST_ENCOUNTERS_VIRAL,
-    ),
     Method(
         "positiveTestsViral_totalTestsViral",
         CommonFields.POSITIVE_TESTS_VIRAL,
         CommonFields.TOTAL_TESTS_VIRAL,
     ),
-    Method(
-        "positiveCasesViral_totalTestsViral",
-        CommonFields.POSITIVE_CASES_VIRAL,
-        CommonFields.TOTAL_TESTS_VIRAL,
-    ),
-    Method(
-        "positiveTests_totalTestsViral", CommonFields.POSITIVE_TESTS, CommonFields.TOTAL_TESTS_VIRAL
-    ),
-    Method(
-        "positiveCasesViral_totalTestsPeopleViral",
-        CommonFields.POSITIVE_CASES_VIRAL,
-        CommonFields.TOTAL_TESTS_PEOPLE_VIRAL,
-    ),
-    Method(
-        "positiveCasesViral_totalTestResults",
-        CommonFields.POSITIVE_CASES_VIRAL,
-        CommonFields.TOTAL_TESTS,
-    ),
 )
 
 
+class TestPositivityException(Exception):
+    pass
+
+
+class NoColumnsWithDataException(TestPositivityException):
+    pass
+
+
+class NoMethodsWithRelevantColumns(TestPositivityException):
+    pass
+
+
+class NoRealTimeseriesValuesException(TestPositivityException):
+    pass
+
+
 @dataclasses.dataclass
 class AllMethods:
     """The result of calculating all test positivity methods for all regions"""
 
     # Test positivity calculated in all valid methods for each region
     all_methods_timeseries: pd.DataFrame
 
-    # Test positivity using the best available method for each region
+    # A MultiRegionTimeseriesDataset with exactly one column, TEST_POSITIVITY, the best available
+    # method for each region.
     test_positivity: MultiRegionTimeseriesDataset
 
     @staticmethod
     def run(
-        metrics_in: MultiRegionTimeseriesDataset,
+        dataset_in: MultiRegionTimeseriesDataset,
         methods: Sequence[Method] = TEST_POSITIVITY_METHODS,
         diff_days: int = 7,
         recent_days: int = 14,
     ) -> "AllMethods":
-        ts_value_cols = list(
-            set(chain.from_iterable((method.numerator, method.denominator) for method in methods))
+        relevant_columns = AllMethods._list_columns(
+            AllMethods._methods_with_columns_available(methods, dataset_in.data.columns)
         )
-        missing_columns = set(ts_value_cols) - set(metrics_in.data.columns)
-        if missing_columns:
-            raise AssertionError(f"Data missing for test positivity: {missing_columns}")
+        if not relevant_columns:
+            raise NoMethodsWithRelevantColumns()
 
-        input_long = metrics_in.timeseries_long(ts_value_cols).set_index(
+        input_long = dataset_in.timeseries_long(relevant_columns).set_index(
             [PdFields.VARIABLE, CommonFields.LOCATION_ID, CommonFields.DATE]
         )[PdFields.VALUE]
         dates = input_long.index.get_level_values(CommonFields.DATE)
+        if dates.empty:
+            raise NoRealTimeseriesValuesException()
```

Comment:
Hrm...  But it shouldn't be an error to run the API pipeline on regions that don't have test positivity data, right?  Or is this an intentional "using exceptions for control flow" type of situation?

Reply:
I see it as a bit gray. It seems like something usually worth warning about. The test_positivity code could log and return a non-error value or raise an exception to let the caller handle it.

Reply:
Okay. I don't care to push back too strongly, but I would consider running the pipeline on a single FIPS to be a normal use case (for testing / development in particular) and we shouldn't even warn about not having data for some columns (all of the test positivity ones), especially if we wouldn't end up warning about it when you run the pipeline on all FIPS.  (i.e. it doesn't seem like you should get a warning about data for FIPS xyz if you run it in isolation, but not if you run it with other FIPS that happen to have more data)